dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala

```diff
@@ -64,40 +66,43 @@ case class Table(
 
-  def putItem[E <: AnyRef](entity: E)(implicit dynamoDB: DynamoDB): Unit = {
-    val attrs = getAttrValuesToMap(entity, true)
+  def putItem[T: u.TypeTag](entity:T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations = getterAnnotationsFromEntity(entity)
+    val fields = getterNamesFromEntity(entity)
 
-    if (attrs("keys").exists(x => x._1 == "hashPK") && attrs("keys").exists(x => x._1 == "rangePK")) {
-      val hashPK = attrs("keys").find(f => f._1 == "hashPK").get._2
-      val rangePK = attrs("keys").find(f => f._1 == "rangePK").get._2
-      dynamoDB.put(this, hashPK, rangePK, attrs("attributes"): _*)
-    } else if (attrs("keys").exists(x => x._1 == "hashPK")) {
-      val hashPK = attrs("keys").find(f => f._1 == "hashPK").get._2
-      dynamoDB.put(this, hashPK, attrs("attributes"): _*)
-    }
-  }
+    val faJoined = annotations.map(tup1 => fields.filter(tup2 => tup2._1 == tup1._1.toString)
+      .map(tup2 => (tup1._1.toString, tup1._2.head.toString, {
+        tup1._3.toString match {
+          case "Int" => tup2._2.asInstanceOf[Int]
+          case "String" => tup2._2.asInstanceOf[String]
+        }
+      }, tup1._3.toString)))
+    val faJoinedFlat = faJoined.flatten
 
-  def putItem[E <: AnyRef](hashPK: Any, rangePK: Any, entity: E)(implicit dynamoDB: DynamoDB): Unit = {
-    val attrs = getAttrValuesToMap(entity, false)
-    dynamoDB.put(this, hashPK, rangePK, attrs("attributes"): _*)
-  }
+    val hashKey = faJoinedFlat.filter(x => x._2.contains("hashPK")).map(y=>y._3).headOption
+    val rangeKey = faJoinedFlat.filter(x => x._2.contains("rangePK")).map(y=>y._3).headOption
 
-  private def getAttrValuesToMap(entity: Any, keysRequired: Boolean): Map[String, List[(String, AnyRef)]] = {
-    val fields = getterNamesFromEntity(entity).map(getterName => {
-      val value = entity.getClass.getDeclaredMethod(getterName).invoke(entity)
-      getterName -> value
-    }).toList
+    val attributes = fields.filter(f => !faJoinedFlat.exists(_._1 == f._1))
+    if(hashKey.isEmpty)
+      throw new Exception(s"Primary key is not defined for ${entity.getClass.getName}")
 
-    val keys = fields.filter(f => f._1 == "hashPK" || f._1 == "rangePK")
-    val attrs = fields.filterNot(f => keys.exists(k => f._1 == k._1))
+    if(hashKey.isDefined && rangeKey.isDefined)
+      dynamoDB.put(this, hashKey.get, rangeKey.get, attributes: _*)
+    else if(hashKey.isDefined)
+      dynamoDB.put(this, hashKey.get, attributes: _*)
+  }
 
-    if (!keys.exists(k => k._1 == "hashPK") && keysRequired)
-      throw new Exception(s"Primary key is not defined for ${entity.getClass.getName}")
+  def putItem[E <: AnyRef](hashPK: Any, rangePK: Any, entity: E)(implicit dynamoDB: DynamoDB): Unit = {
+    val attrs = getterNamesFromEntity(entity)
+    dynamoDB.put(this, hashPK, rangePK, attrs: _*)
+  }
 
-    Map("keys" -> keys, "attributes" -> attrs)
+  private def getterAnnotationsFromEntity[T: u.TypeTag](entity: T) = {
```

Comment:
Let's be more conscious about the return type. Avoid handy tuples as much as possible for readability. Having a private class for the return type is also an option. Having a name for each field in a class is helpful for understanding the meaning of objects here.

Reply:
I have added `AnnotationMeta` case class. This seems to be more consistent now.